engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java

```diff
@@ -158,20 +160,29 @@ private Type getSupertype() {
   }
 
   public final Type[] allTypes(EnsoContext ctx) {
-    if (supertype == null) {
-      if (builtin) {
-        return new Type[] {this};
+    var types = new Type[3];
+    var realCount = fillInTypes(this, types, ctx);
+    return Arrays.copyOf(types, realCount);
+  }
+
+  @ExplodeLoop
+  private static int fillInTypes(Type self, Type[] fill, EnsoContext ctx) {
+    var at = 0;
+    for (; ; ) {
+      fill[at++] = self;
+      if (self.supertype == null) {
+        if (self.builtin) {
+          return at;
+        }
+        fill[at++] = ctx.getBuiltins().any();
+        return at;
```

Comment:
I assume that this gives better performance, but the new code is much harder to follow for me. Is there any chance that we could get good performance with somehow simpler code?

I assume overall we have the following 3 cases:
1. Builtins without a parent
2. Regular types with parent `Any` type
3. Numeric types with `Integer <: Number <: Any`

Am I right? Could we maybe create 3 functions, one for each case?

If that's not possible, I'd at least appreciate a comment explaining what 'paths' this current code is supposed to be generating, so that one could at least verify if the behaviour is as intended.

Reply:
> Am I right?

Yes, you are right. I also think that a little refactoring will not hurt. It is difficult to follow the code without this knowledge. Maybe, it does not even have to be a loop at the end, and just three if statements?

Reply:
Added some javadoc in fe6d2be